node/impl/client/client.go

```diff
@@ -259,6 +259,10 @@ func (a *API) ClientImport(ctx context.Context, ref api.FileRef) (cid.Cid, error
 	return nd, nil
 }
 
+func (a *API) ClientDelete(ctx context.Context, dataCid cid.Cid) error {
+	return a.LocalDAG.Remove(ctx, dataCid)
```

Comment:
This only removes the root node, we need to remove the rest of the DAG too (but not remove blocks shared with other files..)

Reply:
@magik6k 

> but not remove blocks shared with other files

Hrm, that's bad. 

Do there exist any tools in the `DAGService` world which you can point me to which would help with this?

Reply:
So we could just traverse the DAG and remove all nodes, but - and this is where it gets really tricky - we need a way to only remove nodes that aren't shared with other files.

So those are the choices we have:
1. Use the [pinner](https://github.com/ipfs/go-ipfs-pinner) and [gc](https://github.com/ipfs/go-ipfs/blob/master/gc/gc.go) from go-ipfs
  * Pin file roots on import, unpin+gc on delete
  * Not very easy to implement, but doable
2. Remove client import / client blockstore. Instead of passing a CID to `client deal` pass a file path
  * Not super hard to implement
    * Create a temporary per deal filestore/blockstore, drop it after the data got to the miner
    * (To avoid having to make any changes to markets/Data Transfer module, keep a virtual root client blockstore unifying these temporary blockstores (and the optional ipfs bridge blockstore))
  * If we don't want to bother automatically detecting when data transfer for a deal is done, we can just keep the data around and have a separate command to handle deleting that data
    * Bonus points if the once imported data can be reused for other deals
  * IMO better UX - 1 thing to call instead of 2
3. Just embed a full go-ipfs node in lotus to handle client stuff
  * lol

I think 2. is the optimal thing to do, but is fairly involved

Reply:
I lean towards having a separate blockstore per deal, it seems to make it the easiest to reason about cleaning things up later.

Reply:
Nice, good stuff. Thanks.